lib/parallel/cmd_controller.go

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package parallel
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+)
+
+var errCmdFailed = errors.New("command failed")
+
+type cmdController struct {
+	Cmd          Cmd
+	EventHandler func(*Event)
+	Clock        func() time.Time
+	Started      bool
+	Finished     bool
+	StartTime    time.Time
+	Lock         sync.Mutex
+}
+
+func newCmdController(cmd Cmd, eventHandler func(*Event), clock func() time.Time) *cmdController {
+	return &cmdController{cmd, eventHandler, clock, false, false, clock(), sync.Mutex{}}
+}
+
+// Run returns false on failure that has not been already handled
+func (c *cmdController) Run() bool {
+	c.Lock.Lock()
```

Comment:
It's not actually exported since the struct is private, this is a pattern I've been using recently - if you later change the struct to public, it makes it easier.